### PR TITLE
forceDelete() models that use the SoftDeletes trait

### PR DESCRIPTION
**Pull request by jochensengier**
Would be great if this got merged, but as always: your package -> your choice.

Thx!

### CODE REVIEW

File: src/Commands/CleanUpModelsCommand.php [hunk 1]
```diff
@@ -76,7 +76,7 @@ protected function cleanUp(GetsCleanedUp $model)
                 $query->limit($cleanupConfig->chunkBy);
             }
 
-            $numberOfDeletedRecords = $query->delete();
+            $numberOfDeletedRecords = (in_array('Illuminate\Database\Eloquent\SoftDeletes', class_uses($model)) ? $query->forceDelete() : $query->delete());
```
Comment (freekmurze) [inline src/Commands/CleanUpModelsCommand.php]: Use `::class` notation here.

**Review by freekmurze — CHANGES_REQUESTED** [review]
Could you also write a test?

**Comment by jochensengier** [comment]
I used the ::class notation and made some changes for the tests.
Please triple-check and have mercy, I'm not an expert like you in writing tests... 🙈

File: tests/Models/TestSoftDeletableModel.php [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace Tests\Models;
+
+use Closure;
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Database\Eloquent\SoftDeletes;
+use Spatie\ModelCleanup\CleanupConfig;
+use Spatie\ModelCleanup\GetsCleanedUp;
+
+class TestSoftDeletableModel extends Model implements GetsCleanedUp
```
Comment (freekmurze) [inline tests/Models/TestSoftDeletableModel.php]: I'm thinking we could extend the other test model. This way we only have to add `SoftDeletes` and set the table name.

File: tests/TestModelFactory.php [hunk 3]
```diff
@@ -35,7 +36,9 @@ public function create()
         $createdAt = $this->startingFrom;
 
         foreach (range(1, $this->numberOfDays) as $i) {
-            TestModel::create(['created_at' => $createdAt->subDay()]);
+            $date = $createdAt->subDay();
```
Comment (freekmurze) [inline tests/TestModelFactory.php]: Let's create a separate factory for this.

**Review by freekmurze — CHANGES_REQUESTED** [review]
Right now, all tests are modified. 

I'd rater have one added test that proves that it uses soft deletes instead of delete for the right models.

So maybe you could have a tests where you seed models, call delete  on some of them (so they get marked as deleted), and then run the clean up command. You should verify if the models marked for deleted where now actually deleted.

File: tests/Helpers.php [hunk 4]
```diff
@@ -12,12 +12,12 @@
 use function PHPUnit\Framework\assertTrue;
 use Tests\Models\TestModel;
 
-function useCleanupConfig(Closure $closure)
+function useCleanupConfig(Closure $closure, $model)
```
Comment (freekmurze) [inline tests/Helpers.php]: Please add a typehint

File: tests/CleanupSoftDeletableTest.php [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    ], TestSoftDeletableModel::class);
+
+    TestSoftDeletableModel::whereDate('created_at', '<', '2020-01-01')->delete();
+
+    assertDeletedModelsExistForDays([
+        '2019-12-31',
+        '2019-12-30',
+    ], TestSoftDeletableModel::class);
+
+    artisan(CleanUpModelsCommand::class)->assertExitCode(0);
```
Comment (freekmurze) [inline tests/CleanupSoftDeletableTest.php]: Why is this command run here?

**Review by freekmurze — CHANGES_REQUESTED** [review]

**Comment by jochensengier** [comment]
Would you be so kind to have another look at my changes?

File: tests/TestSoftDeletableModelFactory.php [hunk 6]
```diff
@@ -0,0 +1,41 @@
+<?php
+
+namespace Tests;
+
+use Carbon\Carbon;
+use Tests\Models\TestSoftDeletableModel;
+
+class TestSoftDeletableModelFactory
```
Comment (freekmurze) [inline tests/TestSoftDeletableModelFactory.php]: Can you let this factory extend the other factory? This way a lot of code here can be removed.

**Review by freekmurze — CHANGES_REQUESTED** [review]

File: tests/TestSoftDeletableModelFactory.php (hunk 6 above)
Comment (jochensengier) [reply tests/TestSoftDeletableModelFactory.php]: done

**Comment by freekmurze** [comment]
Code looks good! 

Final thing: could you update the readme an add a section on how it behaviour for models with soft deletes?

**Comment by freekmurze** [comment]
Thanks!

**Comment by jochensengier** [comment]
Thanks for accepting the PR and for your patience while I struggled with writing the test.
I couldn't have done it without your guidance.

**Comment by freekmurze** [comment]
Thanks for your work on this.

**Pull request merged**